precompiles/utils/src/data.rs

```diff
@@ -0,0 +1,320 @@
+		T::read(self)
+	}
+
+	/// Parse (4 bytes) selector.
+	/// Returns an error if trying to parse out of bounds.
+	pub fn read_selector(&mut self) -> EvmResult<&[u8]> {
+		let range_end = self.cursor + 4;
+
+		let data = self
+			.input
+			.get(self.cursor..range_end)
+			.ok_or_else(|| error("tried to parse selector out of bounds"))?;
+
+		self.cursor += 4;
+		self.update_max_read_position(self.cursor);
+
+		Ok(data)
+	}
+
+	fn update_max_read_position(&mut self, local_max: usize) {
+		if self.max_read_position < local_max {
+			self.max_read_position = local_max;
+		}
+	}
+}
+
+/// Help build an EVM input/output data.
+#[derive(Clone, Debug)]
+pub struct EvmDataWriter {
+	data: Vec<u8>,
+	arrays: Vec<Array>,
+}
+
+#[derive(Clone, Debug)]
+struct Array {
+	offset_position: usize,
+	data: Vec<u8>,
+	inner_arrays: Vec<Array>,
+}
+
+impl EvmDataWriter {
+	/// Creates a new empty output builder.
+	pub fn new() -> Self {
+		Self {
+			data: vec![],
+			arrays: vec![],
+		}
+	}
+
+	/// Return the built data.
+	pub fn build(mut self) -> Vec<u8> {
+		Self::build_arrays(&mut self.data, self.arrays, 0);
+
+		self.data
+	}
+
+	/// Build the array into data.
+	/// `global_offset` represents the start of the frame we are modifying.
+	/// While the main data will have a `global_offset` of 0, inner arrays will have a
+	/// `global_offset` corresponding to the start its parent array size data.
+	fn build_arrays(output: &mut Vec<u8>, arrays: Vec<Array>, global_offset: usize) {
+		for mut array in arrays {
+			let offset_position = array.offset_position;
+			let offset_position_end = offset_position + 32;
+			let free_space_offset = output.len() + global_offset;
+
+			// Override dummy offset to the offset it will be in the final output.
+			U256::from(free_space_offset)
+				.to_big_endian(&mut output[offset_position..offset_position_end]);
+
+			// Build inner arrays if any.
+			Self::build_arrays(&mut array.data, array.inner_arrays, free_space_offset);
+
+			// Append this data at the end of the current output.
+			output.append(&mut array.data);
+		}
+	}
+
+	/// Write arbitrary bytes.
+	pub fn write_raw_bytes(mut self, value: &[u8]) -> Self {
+		self.data.extend_from_slice(value);
+		self
+	}
+
+	/// Write data of requested type.
+	pub fn write<T: EvmData>(mut self, value: T) -> Self {
+		T::write(&mut self, value);
+		self
+	}
+}
+
+impl Default for EvmDataWriter {
+	fn default() -> Self {
+		Self::new()
+	}
+}
+
+/// Data that can be converted from and to EVM data types.
+pub trait EvmData: Sized {
+	fn read(reader: &mut EvmDataReader) -> EvmResult<Self>;
+	fn write(writer: &mut EvmDataWriter, value: Self);
+}
+
+impl EvmData for H256 {
+	fn read(reader: &mut EvmDataReader) -> EvmResult<Self> {
+		let range_end = reader.cursor + 32;
+
+		let data = reader
+			.input
+			.get(reader.cursor..range_end)
+			.ok_or_else(|| error("tried to parse H256 out of bounds"))?;
+
+		reader.cursor += 32;
+		reader.update_max_read_position(reader.cursor);
```

Comment:
Seems like these two lines happen together a lot. I wonder if the function should be called `update_cursor` or something similar and take care of both parts. I'm not totally clear what the max_read_position` is for though.

Reply:
I moved the cursor movement into a `move_cursor` function that perform overflow checks, updates the cursor and return the range used to read data. I removed `max_read_position` and `check_complete` as additional data is not really an issue, and checking correct data length in tests can be done more simply.